proxy/src/main/java/com/wavefront/agent/TenantInfo.java

```diff
@@ -0,0 +1,196 @@
+import static java.lang.String.format;
+
+/**
+ * The class for keeping tenant required information token, server.
+ *
+ * @author Norayr Chaparyan (nchaparyan@vmware.com).
+ */
+public class TenantInfo {
+    private static final String GET_CSP_ACCESS_TOKEN_ERROR_MESSAGE = "Failed to get access token from CSP.";
+    private static final String CSP_AUTH_TOKEN = "csp-auth-token";
+    // The CSP console url.
+    private static final String cspBaseUrl = "https://console-stg.cloud.vmware.com";
+
+    // This can be a Wavefront API token or a CSP API token.
+    private String tenantToken;
+    private final String tenantServer;
+    private final LoadingCache<String, TokenExchangeResponseDTO> tokenCache;
+
+    TenantInfo(@Nonnull final String tenantToken,
+               @Nonnull final String tenantServer,
+               ProxyConfig.PROXY_AUTH_METHOD proxyAuthMethod) {
+        this.tenantToken = tenantToken;
+        this.tenantServer = tenantServer;
+        if (proxyAuthMethod == CSP_API_TOKEN) {
+            tokenCache = Caffeine.newBuilder().
+                    expireAfter(new Expiry<String, TokenExchangeResponseDTO>() {
+                        @Override
+                        public long expireAfterCreate(@Nonnull String authToken,
+                                                      @Nonnull TokenExchangeResponseDTO tokenExchangeResponseDTO,
+                                                      long currentTime) {
+                            return TimeUnit.SECONDS.toNanos(tokenExchangeResponseDTO.getExpiresIn());
+                        }
+
+                        @Override
+                        public long expireAfterUpdate(@Nonnull String authToken, @Nonnull TokenExchangeResponseDTO tokenExchangeResponseDTO,
+                                                      long currentTime, long currentDuration) {
+                            return currentDuration;
+                        }
+
+                        @Override
+                        public long expireAfterRead(@Nonnull String authToken, @Nonnull TokenExchangeResponseDTO tokenExchangeResponseDTO,
+                                                    long currentTime, long currentDuration) {
+                            return currentDuration;
+                        }
+                    }).build(key -> getAccessTokenByAPIToken());
+        } else {
+            // This step assumes that the authentication mechanism is WAVEFRONT_API_TOKEN.
+            tokenCache = null;
+        }
+    }
+
+    TenantInfo(@Nonnull final String serverToServiceClientId,
+               @Nonnull final String serverToServiceClientSecret,
+               @Nonnull final String orgId,
+               @Nonnull final String tenantServer) {
+        this.tenantServer = tenantServer;
+        this.tokenCache = Caffeine.newBuilder().
+                expireAfter(new Expiry<String, TokenExchangeResponseDTO>() {
+                    @Override
```

Comment:
looks like duplicated code

Reply:
> looks like duplicated code

No, there are two ways to build a tokencache: using getAccessTokenByAPIToken() or getAccessTokenByClientCredentials().

Reply:
> yeah but you have 25 lines duplicated, and only change 1

ahh got it, i'll make changes